### ISSUE TITLE
associationTest fold change cut-off

### ISSUE THREAD

**Issue by koenvandenberge** [statOmics/tradeSeq#75]
When testing against a fold change cut-off using the `associationTest`, we are checking whether the fold change between subsequent knots meets the set threshold. However, this can result in genes with low counts to be found as the top genes, as these are more likely to in-/decrease with these high fold changes over the short timespan that occurs between 2 subsequent knots. 

A better approach that may avoid this, at least partially, would be to check whether the average expression has changed by at least the required fold change amount within that lineage. It seems to me that this would amount to testing all pairwise combinations of knots against each other in the testing (instead of just the subsequent knots).

That contrast matrix will not be full rank and, since we have a fold change threshold, it seems that which subset of independent contrasts that's being chosen here could matter quite significantly.

**Comment by HectorRDB** [comment]
A few points:
+ The consecutive difference result in a contrast matrice C. The contrast matrix resulting from all pairwise differences can be defined by zeros and linear combinations of columns of C. So the full rank restriction is already C, which is why we picked that original choice? 
+ The problem in our fold change cutt-off is that it makes no differences for between consecutive differences of 1, -1, 1, -1 and 1, 1, 1, 1 (all else being equal). 
+ We do not need to focus on consecutive differences, as long as we have k-1 of them and that all the terms are present. *Long term* interactions might result in more meaningful log fold changes. One example of an alternate design would be to contrast the value at knot 1 with all other knots. That would put too much reliance on the estimation of the first knot but it would be cool to check how that affect the ranking of p-values. 
+ In essence, the association test is testing whether the first derivative is different from zero, using the knots values as interpolation for the derivative right?  It could be possible to change it to ressemble more of a patternTest style.  I am unsure how much that will change this issue but it could be interesting.

**Comment by koenvandenberge** [comment]
Thanks very much!

 - I agree that this motivated our original choice of the contrast matrix `C` when we had not yet introduced the fold change cut-off. The `k-1` subsequent comparisons are full rank and all we needed.
- The sign of the contrasts actually do make a difference, which I think makes sense (see example below).
- Yes! And this is where I think the choice becomes very difficult. As you also say, there are many combinations of such `k-1` contrasts that are independent and involve all lineages. Which one you pick doesn't matter when there's no fold change cut-off. But as soon as you introduce the cut-off it matters a lot (see example).

Here is an example gene in John's OE dataset. The smoothers look as follows:
![image](https://user-images.githubusercontent.com/10134844/94970807-54ea4700-04ba-11eb-830d-710955a8bdc4.png)

And we focus on lineage 1 which has 7 knots. Here are the coefficients and variance-covariance matrix:

```
beta <- c(-1.0225962, -1.2397080, -1.1019727, -0.8609921, -0.3385614, -0.3264623, -1.8066244)
Sigma <- matrix(c(1.601427, 1.595840, 1.596526, 1.596448, 1.596352, 1.596359, 1.596357,
                  1.595840, 1.598979, 1.596530, 1.596202, 1.596349, 1.596373, 1.596361,
                  1.596526, 1.596530, 1.600218, 1.598507, 1.596165, 1.596067, 1.596227,
                  1.596448, 1.596202, 1.598507, 1.600457, 1.596377, 1.595518, 1.595990,
                  1.596352, 1.596349, 1.596165, 1.596377, 1.599540, 1.598141, 1.597084,
                  1.596359, 1.596373, 1.596067, 1.595518, 1.598141, 1.598296, 1.595843,
                  1.596357, 1.596361, 1.596227, 1.595990, 1.597084, 1.595843, 1.601241),
                nrow = 7, ncol=7, byrow=FALSE)

```

Comparing the subsequent knots implementation with the first vs all knots implementation, both return exactly the same Wald test statistic of `384.9972`:

```
## current implementation: compare subsequent knots
Lss <- matrix(0, nrow=7, ncol=6)
rownames(Lss) <- paste0("knot", 1:7)
for(kk in 1:6){
  Lss[kk:(kk+1),kk] <- c(-1, 1)
}
qr(Lss)$rank #full rank

## first knot vs others: also full rank
L1 <- matrix(0, nrow=7, ncol=6)
rownames(L1) <- paste0("knot", 1:7)
for(kk in 1:6){
  L1[c(1, kk+1),kk] <- c(-1, 1)
}
qr(L1)$rank #full rank



# log fold change of subsequent implementation
logFcSS <- t(Lss) %*% beta
logFcSS
# log fold change of first knot implementation
logFC1 <- t(L1) %*% beta
logFC1

# Wald test statistic is the same.
t(logFcSS) %*% solve(t(Lss) %*% Sigma %*% Lss) %*% logFcSS
t(logFC1) %*% solve(t(L1) %*% Sigma %*% L1) %*% logFC1
```

However, when we introduce a fold change cut-off, the implementation matters as the fold changes will be different.

```
# now we use a fold change cut-off
cutoff <- log(1.5)
estSS <- matrix(sign(logFcSS) * (pmax(0, abs(logFcSS) - cutoff)), ncol = 1) # zero or remainder
estSS
est1 <- matrix(sign(logFC1) * (pmax(0, abs(logFC1) - cutoff)), ncol = 1) # zero or remainder
est1

# Wald test statistic is pretty different depending on the implementation.
t(estSS) %*% solve(t(Lss) %*% Sigma %*% Lss) %*% estSS
t(est1) %*% solve(t(L1) %*% Sigma %*% L1) %*% est1
```

To respond to your second point, the sign of the contrasts does seem to matter:

```
## sign of contrasts matters: test statistic is different
Lss[Lss != 0] <- 1
logFcSS <- t(Lss) %*% beta
t(logFcSS) %*% solve(t(Lss) %*% Sigma %*% Lss) %*% logFcSS
```

**Comment by HectorRDB** [comment]
@koenvandenberge ,
I think you forgot to paste the output of the code but I get the gist of it. 

However, the logfc cutoff implementation, as it is currently implemented, is knot-dependant. For example, with a high enough number of knots, no gene will pass the FC cutoff for the `associationTest` since all changes between knots will be too small.  This is problematic I think. Basically, we could have a bandwidth parameter (a function of nkots by default), which will decide the size of the appropriate contrast. If `nknots=3`, consecutive knots might make sense but as more knots are added, we could do longer range interactions. 

Moreover, this might help the related issue of correlation. Further away knots will be less correlated, making the covariance matrix easier to invert. 

What do you think?

**Comment by koenvandenberge** [comment]
Yes, I totally agree that the knot dependence is also an issue that is very related.
As you also said, it may be better to have an `nPoints` argument much like `patternTest`. As a first solution to try out, I was wondering if we should compare the end point with the other selected points. The end point would be more stable than the starting point, as that's typically pretty variable because of the likely random assignment of cells at the start of a trajectory, if there are multiple lineages.

**Comment by HectorRDB** [comment]
I would disagree that the endpoints should be used for a few reasons.

+ While the start point might be less stable to random assignments, their pseudotime value is more stable since the lineages are estimated using more cells.
+ Lineages with few cells at the end will also lead to less stable estimates of the endpoints. On the other hand, start point values are estimated using an equal number of cells per lineage (more or less).
+ This might make the logFC cutoff more meaningful between lineages since start points should be roughly the same so the estimates are more comparable.

**Comment by koenvandenberge** [comment]
After discussion we decided to

 - Add a warning in `fitGAM` if multiple conditions are present to say it's still experimental
 - Add an argument to `associationTest` which could be either `"start"` or `"end"`, defining the reference point for the contrast. The contrast will be the start or end point versus other points.
 - The other points will no longer be the knots but we'll have an `nPoints` argument such that it's no longer knot dependent.

**Comment by HectorRDB** [comment]
Note that `edgeR` had a different solution to this problem: when testing against a threshold, only a simple ttt/ctl threshold is possible (see the `glmTreat` function [help](https://rdrr.io/bioc/edgeR/man/glmTreat.html)). 
In that setting, logFc would only make sense for the `startVSEndTest` and the `diffEndTest`. 
I'm not saying that we should do this. However, this shows that 1/ it is indeed a tricky issue, and 2/ we might want to make the documentation more explicit?

**Comment by koenvandenberge** [comment]
I just finished updating the associationTest and have implemented the points we discussed above.
There are two new arguments:

 - `nPoints` to define the number of points per lineage you'd like to compare.
 - `contrastType` to define the contrast type you'd want to use. It has three options:
   1. `"start"` where each point is compared to the start of each lineage.
   2. `"end"` where each point is compared to the end of the lineage.
   3. `"consecutive"` which is the traditional way it was implemented, comparing consecutive points along the lineage. It is now implemented using points along the lineage as compared to comparing the different knot points.
